Source/Core/RequestScheduler.js

```diff
@@ -0,0 +1,383 @@
+    };
+
+    /**
+     * For testing only.
+     *
+     * @private
+     */
+    RequestScheduler.numberOfActiveRequestsByServer = function(serverName) {
+        return numberOfActiveRequestsByServer[serverName];
+    };
+
+    function serverHasOpenSlots(server) {
+        return numberOfActiveRequestsByServer[server] < RequestScheduler.maximumRequestsPerServer;
+    }
+
+    function issueRequest(request) {
+        if (request.state === RequestState.UNISSUED) {
+            request.state = RequestState.ISSUED;
+            request.deferred = when.defer();
+        }
+        return request.deferred.promise;
+    }
+
+    function getRequestReceivedFunction(request) {
+        return function(results) {
+            if (request.state === RequestState.CANCELLED) {
+                return;
+            }
+            --statistics.numberOfActiveRequests;
+            --numberOfActiveRequestsByServer[request.server];
+            request.state = RequestState.RECEIVED;
+            request.deferred.resolve(results);
+        };
+    }
+
+    function getRequestFailedFunction(request) {
+        return function(error) {
+            if (request.state === RequestState.CANCELLED) {
+                return;
+            }
+            ++statistics.numberOfFailedRequests;
+            --statistics.numberOfActiveRequests;
+            --numberOfActiveRequestsByServer[request.server];
+            request.state = RequestState.FAILED;
+            request.deferred.reject(error);
+        };
+    }
+
+    function startRequest(request, passthrough) {
+        if (passthrough) {
+            // Data uri or blob uri should always succeed so make the request immediately and don't contribute to statistics.
+            request.state = RequestState.RECEIVED;
+            return request.requestFunction();
+        }
+
+        var promise = issueRequest(request);
+        request.state = RequestState.ACTIVE;
+        activeRequests.push(request);
+        ++statistics.numberOfActiveRequests;
+        ++statistics.numberOfActiveRequestsEver;
+        ++numberOfActiveRequestsByServer[request.server];
+
+        request.requestFunction().then(getRequestReceivedFunction(request)).otherwise(getRequestFailedFunction(request));
+        return promise;
+    }
+
+    function cancelRequest(request) {
+        if (request.state === RequestState.ACTIVE) {
+            --statistics.numberOfActiveRequests;
+            --numberOfActiveRequestsByServer[request.server];
+            ++statistics.numberOfCancelledActiveRequests;
+            if (defined(request.xhr)) {
+                // TODO : make sure this doesn't trigger a failed promise, if so the deferred can be rejected first
+                // TODO : test this in a test
+                request.xhr.abort();
+            }
+        }
+        request.state = RequestState.CANCELLED;
+        ++statistics.numberOfCancelledRequests;
+        request.deferred.reject('Cancelled');
```

Comment:
I'll also add that Promise cancellation is a thing that exists and is in Bluebird.  When we finally get around to switching Cesium over, we may be able to use that here instead of rejecting.